README.rst
python-leaderboard
==================

This repository has moved and all future development will be done at https://github.com/agoragames/leaderboard-python. We 
will be deleting this repository within the next 60 days.

CHANGELOG

master
======

1.1.8 (2015-04-02)
==================
Fix Redis connection passing in options, [#3](https://github.com/agoragames/python-leaderboard/pull/3)

1.1.7
=====

Fixes #2, https://github.com/agoragames/python-leaderboard/pull/2. Fixed a
bug where withscores was passed incorrectly on ascending calls.

Changing redis requirements to ensure that ConnectionPool accepts a host
keyword argument

1.1.3
=====

`leaders` will return an empty page if the page number is invalid

1.1.2
=====

Member identifiers are cast to strings where appropriate


1.1.1
=====

Uses connection pooling for each unique (host,port,db) connection

Renamed `add_member` to `rank_member` to clarify that it applies to both
new and existing members

Added `clear` to delete a leaderboard and its members

Page size can be dynamically set in `total_pages`, `leaders`, and `around_me`
as the `page_size` keyword argument


tests/__init__.py
import unittest
from leaderboard_test_case import LeaderboardTestCase

def all_tests():
    suite = unittest.TestSuite()
    suite.addTest(unittest.makeSuite(LeaderboardTestCase))
    return suite


tests/leaderboard_test_case.py
from redis import Redis, ConnectionPool
import leaderboard
import unittest

class LeaderboardTestCase(unittest.TestCase):
    def setUp(self):
        self.leaderboard = leaderboard.Leaderboard('name')

    def tearDown(self):
        self.leaderboard.redis_connection.flushdb()

    def test_init_with_defaults(self):
        self.assertEquals('name', self.leaderboard.leaderboard_name)
        self.assertEquals( 1, len(self.leaderboard.options) )
        self.assertTrue( isinstance(self.leaderboard.options['connection_pool'],ConnectionPool) )
        self.assertTrue( isinstance(self.leaderboard.redis_connection,Redis) )
        self.assertEquals(self.leaderboard.DEFAULT_PAGE_SIZE, self.leaderboard.page_size)

    def test_init_sets_page_size_to_default_if_set_to_invalid_value(self):
        self.leaderboard = leaderboard.Leaderboard('name', page_size=0)
        self.assertEquals(self.leaderboard.DEFAULT_PAGE_SIZE, self.leaderboard.page_size)

    def test_init_uses_connection_pooling(self):
        lb0 = leaderboard.Leaderboard('lb0', db=0)
        lb1 = leaderboard.Leaderboard('lb1', db=0)
        lb2 = leaderboard.Leaderboard('lb2', db=1)

        self.assertEquals( lb0.redis_connection.connection_pool,
            lb1.redis_connection.connection_pool )
        self.assertNotEquals( lb0.redis_connection.connection_pool,
            lb2.redis_connection.connection_pool )

    def test_rank_member_and_total_members(self):
        self.leaderboard.rank_member('member', 1)
        self.assertEquals(1, self.leaderboard.total_members())

    def test_remove_member(self):
        self._add_members_to_leaderboard(self.leaderboard.DEFAULT_PAGE_SIZE)

        self.assertEquals(self.leaderboard.DEFAULT_PAGE_SIZE, self.leaderboard.total_members())

        self.leaderboard.remove_member('member_1')
        self.assertEquals(self.leaderboard.DEFAULT_PAGE_SIZE - 1, self.leaderboard.total_members())
        self.assertEquals(None, self.leaderboard.rank_for('member_1'))

    def test_clear(self):
        self._add_members_to_leaderboard(3)
        self.assertEquals(3, self.leaderboard.total_members() )
        self.leaderboard.clear()
        self.assertFalse( self.leaderboard.redis_connection.exists(
            self.leaderboard.leaderboard_name) )

    def test_total_members_in_score_range(self):
        self._add_members_to_leaderboard(5)
        self.assertEquals(3, self.leaderboard.total_members_in_score_range(2, 4))

    def test_rank_for(self):
        self._add_members_to_leaderboard(5)
        self.assertEquals(2, self.leaderboard.rank_for('member_4'))
        self.assertEquals(1, self.leaderboard.rank_for('member_4', True))

    def test_score_for(self):
        self._add_members_to_leaderboard(5)
        self.assertEquals(4, self.leaderboard.score_for('member_4'))

    def test_total_pages(self):
        self._add_members_to_leaderboard(10)
        self.assertEquals(1, self.leaderboard.total_pages())
        self.leaderboard.redis_connection.flushdb()
        self._add_members_to_leaderboard(self.leaderboard.DEFAULT_PAGE_SIZE + 1)
        self.assertEquals(2, self.leaderboard.total_pages())

    def test_leaders(self):
        self._add_members_to_leaderboard(25)
        self.assertEquals(25, self.leaderboard.total_members())

        leaders = self.leaderboard.leaders(1)

        self.assertEquals(25, len(leaders))
        self.assertEquals('member_25', leaders[0]['member'])
        self.assertEquals('member_2', leaders[-2]['member'])
        self.assertEquals('member_1', leaders[-1]['member'])
        self.assertEquals(1, leaders[-1]['score'])

    def test_leaders_ascending(self):
        lb = leaderboard.Leaderboard('asc',order='asc')

        for index in range(1, 11):
            lb.rank_member("member_%d" % index, index)

        leaders = lb.leaders(1)
        self.assertEquals(10, len(leaders))
        self.assertEquals('member_1', leaders[0]['member'])
        self.assertEquals('member_10', leaders[-1]['member'])

    def test_leaders_with_multiple_pages(self):
        self._add_members_to_leaderboard(self.leaderboard.DEFAULT_PAGE_SIZE * 3 + 1)
        self.assertEquals(self.leaderboard.DEFAULT_PAGE_SIZE * 3 + 1, self.leaderboard.total_members())

        leaders = self.leaderboard.leaders(1)
        self.assertEquals(self.leaderboard.page_size, len(leaders))

        leaders = self.leaderboard.leaders(2)
        self.assertEquals(self.leaderboard.page_size, len(leaders))

        leaders = self.leaderboard.leaders(3)
        self.assertEquals(self.leaderboard.page_size, len(leaders))

        leaders = self.leaderboard.leaders(4)
        self.assertEquals(1, len(leaders))

        leaders = self.leaderboard.leaders(-5)
        self.assertEquals(self.leaderboard.page_size, len(leaders))

    # assert we get an empty page back
        leaders = self.leaderboard.leaders(10)
        self.assertEquals(None, leaders)

    def test_around_me(self):
        self._add_members_to_leaderboard(self.leaderboard.DEFAULT_PAGE_SIZE * 3 + 1)
        self.assertEquals(self.leaderboard.DEFAULT_PAGE_SIZE * 3 + 1, self.leaderboard.total_members())

        leaders_around_me = self.leaderboard.around_me('member_30')
        self.assertEquals(self.leaderboard.page_size / 2, len(leaders_around_me) / 2)

        leaders_around_me = self.leaderboard.around_me('member_1')
        self.assertEquals(self.leaderboard.page_size / 2 + 1, len(leaders_around_me))

        leaders_around_me = self.leaderboard.around_me('member_76')
        self.assertEquals(self.leaderboard.page_size / 2, len(leaders_around_me) / 2)

    def test_ranked_in_list(self):
        self._add_members_to_leaderboard(self.leaderboard.DEFAULT_PAGE_SIZE)
        self.assertEquals(self.leaderboard.DEFAULT_PAGE_SIZE, self.leaderboard.total_members())

        members = ['member_1', 'member_5', 'member_10']
        ranked_members = self.leaderboard.ranked_in_list(members, True)

        self.assertEquals(3, len(ranked_members))

        self.assertEquals(25, ranked_members[0]['rank'])
        self.assertEquals(1, ranked_members[0]['score'])

        self.assertEquals(21, ranked_members[1]['rank'])
        self.assertEquals(5, ranked_members[1]['score'])

        self.assertEquals(16, ranked_members[2]['rank'])
        self.assertEquals(10, ranked_members[2]['score'])

    def test_change_score_for(self):
        self.leaderboard.rank_member('member_1', 5)
        self.assertEquals(5, self.leaderboard.score_for('member_1'))

        self.leaderboard.change_score_for('member_1', 5)
        self.assertEquals(10, self.leaderboard.score_for('member_1'))

        self.leaderboard.change_score_for('member_1', -5)
        self.assertEquals(5, self.leaderboard.score_for('member_1'))

    def test_check_member(self):
        self.leaderboard.rank_member('member_1', 10)

        self.assertEquals(True, self.leaderboard.check_member('member_1'))
        self.assertEquals(False, self.leaderboard.check_member('member_2'))

    def test_can_change_page_size_and_have_it_reflected_in_size_of_result_set(self):
        self._add_members_to_leaderboard(self.leaderboard.DEFAULT_PAGE_SIZE)

        self.leaderboard.page_size = 5
        self.assertEquals(5, self.leaderboard.total_pages())

        self.assertEquals(5, len(self.leaderboard.leaders(1)))

    def test_score_and_rank_for(self):
        self._add_members_to_leaderboard()

        data = self.leaderboard.score_and_rank_for('member_1')
        self.assertEquals('member_1', data['member'])
        self.assertEquals(1, data['score'])
        self.assertEquals(5, data['rank'])

    def test_remove_members_in_score_range(self):
        self._add_members_to_leaderboard()

        self.assertEquals(5, self.leaderboard.total_members())

        self.leaderboard.rank_member('cheater_1', 100)
        self.leaderboard.rank_member('cheater_2', 101)
        self.leaderboard.rank_member('cheater_3', 102)

        self.assertEquals(8, self.leaderboard.total_members())

        self.leaderboard.remove_members_in_score_range(100, 102)

        self.assertEquals(5, self.leaderboard.total_members())

        for leader in self.leaderboard.leaders(1):
            self.assertTrue(leader['score'] < 100)

    def test_pipeline(self):
        pipeline = self.leaderboard.pipeline()

        self._add_members_to_leaderboard(3, pipeline)

        self.assertEquals(0, self.leaderboard.total_members())
        self.leaderboard.commit(pipeline)
        self.assertEquals(3, self.leaderboard.total_members())

    def _add_members_to_leaderboard(self, members_to_add = 5, pipeline = None):
        for index in range(1, members_to_add + 1):
            self.leaderboard.rank_member("member_%d" % index, index, pipeline = pipeline)


    def test_redis_connection(self):
      lb = leaderboard.Leaderboard('name', connection=Redis())
      self.assertEquals( 0, len(lb.options) )

leaderboard.py
from redis import Redis
from copy import deepcopy
from math import ceil

class Leaderboard(object):
	VERSION = '1.0.1'
	DEFAULT_PAGE_SIZE = 25
	DEFAULT_REDIS_HOST = 'localhost'
	DEFAULT_REDIS_PORT = 6379

	def __init__(self, leaderboard_name, host = DEFAULT_REDIS_HOST, port = DEFAULT_REDIS_PORT, page_size = DEFAULT_PAGE_SIZE, **redis_options):
		self.leaderboard_name = leaderboard_name
		self.host = host
		self.port = port

		if page_size < 1:
			page_size = self.DEFAULT_PAGE_SIZE

		self.page_size = page_size

		self.redis_options = deepcopy(redis_options)
		self.redis_options.setdefault('host', self.host)
		self.redis_options.setdefault('port', self.port)

		self.redis_connection = Redis(self.redis_options['host'], self.redis_options['port'], self.redis_options.get('db',0))

	def add_member(self, member, score):
		return self.redis_connection.zadd(self.leaderboard_name, member, score)

	def remove_member(self, member):
		return self.redis_connection.zrem(self.leaderboard_name, member)

	def total_members(self):
		return self.redis_connection.zcard(self.leaderboard_name)

	def total_pages(self):
		return ceil(float(self.total_members()) / self.page_size)

	def total_members_in_score_range(self, min_score, max_score):
		return self.redis_connection.zcount(self.leaderboard_name, min_score, max_score)

	def change_score_for(self, member, delta):
		return self.redis_connection.zincrby(self.leaderboard_name, member, delta)

	def rank_for(self, member, use_zero_index_for_rank = False):
		if use_zero_index_for_rank:
			return self.redis_connection.zrevrank(self.leaderboard_name, member)
		else:
			try: return self.redis_connection.zrevrank(self.leaderboard_name, member) + 1
			except: return None

	def score_for(self, member):
		return self.redis_connection.zscore(self.leaderboard_name, member)

	def check_member(self, member):
		return not None == self.redis_connection.zscore(self.leaderboard_name, member)

	def score_and_rank_for(self, member, use_zero_index_for_rank = False):
		return {'member' : member, 'score' : self.score_for(member), 'rank' : self.rank_for(member, use_zero_index_for_rank)}

	def remove_members_in_score_range(self, min_score, max_score):
		return self.redis_connection.zremrangebyscore(self.leaderboard_name, min_score, max_score)

	def leaders(self, current_page, with_scores = True, with_rank = True, use_zero_index_for_rank = False):
		if current_page < 1:
			current_page = 1
			
		tpages = self.total_pages()
			
		if current_page > tpages:
			current_page = tpages

		index_for_redis = current_page - 1

		starting_offset = (index_for_redis * self.page_size)
		if starting_offset < 0:
			starting_offset = 0

		ending_offset = (starting_offset + self.page_size) - 1

		raw_leader_data = self.redis_connection.zrevrange(self.leaderboard_name, int(starting_offset), int(ending_offset), with_scores)
		if raw_leader_data:
			return self._massage_leader_data(raw_leader_data, with_rank, use_zero_index_for_rank)
		else:
			return None

	def around_me(self, member, with_scores = True, with_rank = True, use_zero_index_for_rank = False):
		reverse_rank_for_member = self.redis_connection.zrevrank(self.leaderboard_name, member)

		starting_offset = reverse_rank_for_member - (self.page_size / 2)
		if starting_offset < 0:
			starting_offset = 0

		ending_offset = (starting_offset + self.page_size) - 1

		raw_leader_data = self.redis_connection.zrevrange(self.leaderboard_name, starting_offset, ending_offset, with_scores)
		if raw_leader_data:
			return self._massage_leader_data(raw_leader_data, with_rank, use_zero_index_for_rank)
		else:
			return None

	def ranked_in_list(self, members, with_scores = True, use_zero_index_for_rank = False):
		ranks_for_members = []

		for member in members:
			data = {}
			data['member'] = member
			data['rank'] = self.rank_for(member, use_zero_index_for_rank)
			if with_scores:
				data['score'] = self.score_for(member)

			ranks_for_members.append(data)

		return ranks_for_members

	def _massage_leader_data(self, leaders, with_rank, use_zero_index_for_rank):
		member_attribute = True
		leader_data = []

		for leader_data_item in leaders:
			data = {}
			data['member'] = leader_data_item[0]
			data['score'] = leader_data_item[1]
			if with_rank:
				data['rank'] = self.rank_for(data['member'], use_zero_index_for_rank)

			leader_data.append(data)

		return leader_data

leaderboard/__init__.py
from redis import Redis, ConnectionPool
from copy import deepcopy
from math import ceil


class Leaderboard(object):
    VERSION = '1.1.8'
    DEFAULT_PAGE_SIZE = 25
    DEFAULT_REDIS_HOST = 'localhost'
    DEFAULT_REDIS_PORT = 6379
    DEFAULT_REDIS_DB = 0
    ASC = 'asc'
    DESC = 'desc'

    @classmethod
    def pool(self, host, port, db, pools={}):
        '''
        Fetch a redis conenction pool for the unique combination of host
        and port. Will create a new one if there isn't one already.
        '''
        key = (host,port,db)
        rval = pools.get( key )
        if not isinstance(rval,ConnectionPool):
            rval = ConnectionPool(host=host, port=port, db=db)
            pools[ key ] = rval
        return rval

    def __init__(self, leaderboard_name, **options):
        '''
        Initialize a connection to a specific leaderboard. By default, will use a
        redis connection pool for any unique host:port:db pairing.

        The options and their default values (if any) are:

        host : the host to connect to if creating a new handle ('localhost')
        port : the port to connect to if creating a new handle (6379)
        db : the redis database to connect to if creating a new handle (0)
        page_size : the default number of items to return in each page (25)
        connection : an existing redis handle if re-using for this leaderboard
        connection_pool : redis connection pool to use if creating a new handle
        '''
        self.leaderboard_name = leaderboard_name
        self.redis_connection = options.pop('connection', None)
        self.options = deepcopy(options)

        self.page_size = self.options.pop('page_size', self.DEFAULT_PAGE_SIZE)
        if self.page_size < 1:
            self.page_size = self.DEFAULT_PAGE_SIZE

        self.order = self.options.pop('order', self.DESC).lower()
        if not self.order in [self.ASC, self.DESC]:
            raise ValueError("%s is not one of [%s]" % (self.order,  ",".join([self.ASC, self.DESC])))

        if not isinstance(self.redis_connection, Redis):
            if 'connection_pool' not in self.options:
                self.options['connection_pool'] = self.pool(
                    self.options.pop('host', self.DEFAULT_REDIS_HOST),
                    self.options.pop('port', self.DEFAULT_REDIS_PORT),
                    self.options.pop('db', self.DEFAULT_REDIS_DB)
                )
            self.redis_connection = Redis(**self.options)

    @classmethod
    def get_connection(self, **options):
        redis_options = {}
        redis_options['connection_pool'] = self.pool(
            options.pop('host', self.DEFAULT_REDIS_HOST),
            options.pop('port', self.DEFAULT_REDIS_PORT),
            options.pop('db', self.DEFAULT_REDIS_DB)
        )
        return Redis(**redis_options)

    def _get(self, **options):
        if 'pipeline' in options and options['pipeline'] != None:
            return options['pipeline']
        return self.redis_connection

    def pipeline(self):
        return self.redis_connection.pipeline()

    def commit(self, pipeline):
        pipeline.execute()

    def rank_member(self, member, score, **options):
        # redis-py deprecated the non-kwarg form of zadd
        return self._get(**options).zadd(self.leaderboard_name, **{str(member):score})

    def remove_member(self, member, **options):
        return self._get( **options).zrem(self.leaderboard_name, str(member))

    def clear(self, **options):
        '''Remove all rankings for this leaderboard.'''
        self._get(**options).delete(self.leaderboard_name)

    def total_members(self, **options):
        return self._get(**options).zcard(self.leaderboard_name)

    def total_pages(self, **options):
        return ceil(float(self.total_members(**options)) / options.get('page_size',self.page_size))

    def total_members_in_score_range(self, min_score, max_score, **options):
        return self._get(**options).zcount(self.leaderboard_name, min_score, max_score)

    def change_score_for(self, member, delta, **options):
        return self._get(**options).zincrby(self.leaderboard_name, str(member), delta)

    def rank_for(self, member, use_zero_index_for_rank = False, **options):
        try:
            return self._rank_method(self._get(**options), self.leaderboard_name, str(member))\
                + (0 if use_zero_index_for_rank else 1)
        except: return None

    def score_for(self, member, **options):
        return self._get(**options).zscore(self.leaderboard_name, str(member))

    def check_member(self, member, **options):
        return not None == self._get(**options).zscore(self.leaderboard_name, str(member))

    def score_and_rank_for(self, member, use_zero_index_for_rank = False, **options):
        return {
            'member' : member,
            'score' : self.score_for(member, **options),
            'rank' : self.rank_for(member, use_zero_index_for_rank, **options)
        }

    def remove_members_in_score_range(self, min_score, max_score, **options):
        return self._get(**options).zremrangebyscore(self.leaderboard_name, min_score, max_score)

    def leaders(self, current_page, with_scores = True, with_rank = True, use_zero_index_for_rank = False, **options):
        if current_page < 1:
            current_page = 1

        page_size = options.get('page_size',self.page_size)
        tpages = self.total_pages(page_size=page_size)

        index_for_redis = current_page - 1

        starting_offset = (index_for_redis * page_size)
        if starting_offset < 0:
            starting_offset = 0

        ending_offset = (starting_offset + page_size) - 1

        raw_leader_data = self._range_method(self._get(**options), self.leaderboard_name, int(starting_offset), int(ending_offset), withscores=with_scores)
        if raw_leader_data:
            return self._massage_leader_data(raw_leader_data, with_rank, use_zero_index_for_rank)
        else:
            return None

    def around_me(self, member, with_scores = True, with_rank = True, use_zero_index_for_rank = False, **options):
        reverse_rank_for_member = \
            self._rank_method(self._get(**options), self.leaderboard_name, str(member))

        if not reverse_rank_for_member is None:
            page_size = options.get('page_size',self.page_size)
            starting_offset = reverse_rank_for_member - (page_size / 2)
            if starting_offset < 0:
                starting_offset = 0

            ending_offset = (starting_offset + page_size) - 1

            raw_leader_data = self._range_method(self._get(**options), self.leaderboard_name, starting_offset, ending_offset, withscores=with_scores)
            if raw_leader_data:
                return self._massage_leader_data(raw_leader_data, with_rank, use_zero_index_for_rank)

        return None

    def ranked_in_list(self, members, with_scores = True, use_zero_index_for_rank = False):
        ranks_for_members = []

        for member in members:
            data = {}
            data['member'] = member
            data['rank'] = self.rank_for(member, use_zero_index_for_rank)
            if with_scores:
                data['score'] = self.score_for(member)

            ranks_for_members.append(data)

        return ranks_for_members

    def _range_method(self, connection, *args, **kwargs):
        if self.order == self.DESC:
            return connection.zrevrange(*args, **kwargs)
        else:
            return connection.zrange(*args, **kwargs)

    def _rank_method(self, connection, *args, **kwargs):
        if self.order == self.DESC:
            return connection.zrevrank(*args, **kwargs)
        else:
            return connection.zrank(*args, **kwargs)

    def _massage_leader_data(self, leaders, with_rank, use_zero_index_for_rank):
        member_attribute = True
        leader_data = []

        for leader_data_item in leaders:
            data = {}
            data['member'] = leader_data_item[0]
            data['score'] = leader_data_item[1]
            if with_rank:
                data['rank'] = self.rank_for(data['member'], use_zero_index_for_rank)

            leader_data.append(data)

        return leader_data
